File: practice/src/org/acme/grabber/entities/TunedStringSet.java
package org.acme.grabber.entities;

import lombok.Getter;
import org.acme.grabber.enums.Note;
import org.acme.grabber.enums.StringSet;
import org.acme.grabber.enums.StringSize;
import org.acme.grabber.enums.Tuning;

import java.util.ArrayList;
import java.util.List;

import static org.testng.Assert.assertEquals;

@Getter
public class TunedStringSet {

    private static final String ERROR_MESSAGE = "Tuning/String Set size mismatch. Choose equally sized sets/tunings.";
    private List<GuitarString> stringList = new ArrayList<>();

    public TunedStringSet(Tuning tuning, StringSet stringSet) {
        assertEquals(tuning.getNoteList().size(), stringSet.getStringSizeList().size(), ERROR_MESSAGE);

        for (int i = 0; i < tuning.getNoteList().size(); i++) {
            stringList.add(new GuitarString(tuning.getNoteList().get(i), stringSet.getStringSizeList().get(i)));
        }
    }

    public TunedStringSet addString(Note note, StringSize stringSize) {
        stringList.add(new GuitarString(note, stringSize));
        return this;
    }
}

File: practice/src/org/acme/grabber/entities/TunedStringSet.kt
package org.acme.grabber.entities

import org.acme.grabber.pages.TensionCalcPage
import selenium.WebDriverHolder
import java.util.regex.Pattern

class TunedStringSet(private val tuning: Tuning, private val vendorStringSet: VendorStringSet) {
    private val strings = mutableListOf<GuitarString>()
    private val errorMessage = "Tuning/String Set size mismatch. Choose equally sized sets/tunings."

    init {
        if (tuning.notes.size == vendorStringSet.stringSizes.size) {
            for (i in tuning.notes.indices) {
                strings.add(GuitarString(tuning.notes[i], vendorStringSet.stringSizes[i]))
            }
        } else {
            throw IllegalArgumentException(errorMessage)
        }
    }

    fun calculateTension() {
        WebDriverHolder.webDriver.get(TensionCalcPage.url)
        strings.forEach {
            val fullOutput = TensionCalcPage
                    .setStringSize(it.size.value)
                    .setNoteSize(it.note.value)
                    .clickCalculateButton()
                    .getCalculationOutputText()
            it.tension = extractTensionValue(fullOutput)
        }
    }

    fun prettyPrint() {
        println("Tension chart for ${vendorStringSet.displayName} in ${tuning.displayName} tuning:")
        strings.forEach { println(it.toString()) }
    }

    private fun extractTensionValue(fullOutput: String?): String {
        val p = Pattern.compile("[\\d.]+")
        val m = p.matcher(fullOutput!!)
        return if (m.find()) m.group(0) else "Value not found..."
    }
}
